### Import required dependencies

In [1]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [2]:
# Create the path and import the data
full_student_data = os.path.join('../Resources/new_full_student_data.csv')
student_df = pd.read_csv(full_student_data)

In [3]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [4]:
# Check for null values
student_df.isnull().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [5]:
# Drop rows with null values and verify removal
student_df = student_df.dropna()
student_df.isnull().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [6]:
# Check for duplicated rows
student_df.duplicated().sum()

1836

In [7]:
# Drop duplicated rows and verify removal
student_df = student_df.drop_duplicates()
student_df.duplicated().sum()

0

In [8]:
# Check data types
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [9]:
# Examine the grade column to understand why it is not an int
student_df.head()

student_id      student_name grade             school_name  reading_score  \
0   103880842     Travis Martin   9th    Sullivan High School           59.0   
1    45069750     Michael Brown   9th       Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th      Wagner High School           89.0   
3    62582498  Susan Richardson   9th       Silva High School           69.7   
5    74579444   Cynthia Johnson   9th  Montgomery High School           63.5   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
5        76.9     Charter         893368

In [10]:
# Remove the non-numeric characters and verify the contents of the column
student_df.loc[:, "grade"] = student_df.loc[:, "grade"].str.replace("th", "")
student_df.head()
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [11]:
# Change the grade column to the int type and verify column types
student_df.loc[:, "grade"] = student_df.loc[:, "grade"].astype("int")
student_df.dtypes

student_id         int64
student_name      object
grade              int32
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [12]:
# Display summary statistics for the DataFrame
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [13]:
# Display the mean math score using the mean function
student_df["math_score"].mean()

64.67573326141189

In [14]:
# Store the minimum reading score as min_reading_score
min_reading_score = student_df["reading_score"].min()
min_reading_score

10.5

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [24]:
# Use loc to display the grade column
student_df.loc[:, "grade"]

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: int32

In [16]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
student_df.iloc[0:3, 3:6]

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [17]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
student_df_filter = student_df.loc[student_df["grade"] == 9]
student_df_filter.describe()

student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [37]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_reading_row = student_df.loc[student_df["reading_score"] == min_reading_score]
min_reading_row


student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [19]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
reading_scores_Dixon = student_df.loc[(student_df["grade"] == 10) & (student_df["school_name"] == "Dixon High School")]
reading_scores_Dixon.loc[:, "reading_score"]

45       71.1
60       59.5
69       88.6
94       81.5
100      95.3
         ... 
19283    52.9
19306    58.0
19344    38.0
19368    84.4
19445    43.9
Name: reading_score, Length: 569, dtype: float64

In [20]:
# Find the mean reading score for all students in grades 11 and 12 combined.

avg_reading_score = student_df.loc[(student_df["grade"] == 11) | (student_df["grade"] == 12)]
avg_reading_score.loc[:, "reading_score"].mean()

74.90038089192188

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

In [44]:
# Use groupby and mean to find the average reading and math scores for each school type.
avg_score_by_school = student_df.groupby("school_type").mean()
avg_score_by_school


student_id      grade  reading_score  math_score  school_budget
school_type                                                                   
Charter      7.004029e+07  10.346744      72.450603   66.761883  872625.656236
Public       6.951549e+07  10.362808      72.281219   62.951576  911195.558251

In [58]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
total_students_by_school = student_df.groupby("school_name").count()
total_students_by_school


student_id  student_name  grade  reading_score  \
school_name                                                              
Bowers High School             803           803    803            803   
Campbell High School           407           407    407            407   
Campos High School             541           541    541            541   
Chang High School              171           171    171            171   
Dixon High School             1583          1583   1583           1583   
Fisher High School             798           798    798            798   
Green High School             1961          1961   1961           1961   
Montgomery High School        2038          2038   2038           2038   
Odonnell High School           459           459    459            459   
Richard High School            551           551    551            551   
Silva High School             1109          1109   1109           1109   
Sullivan High School           971           971    971            971   
Turner High School             846           846    846            846   
Wagner High School            1541          1541   1541           1541   
Woods High School             1052          1052   1052           1052   

                        math_score  school_type  school_budget  
school_name                                                     
Bowers High School             803          803            803  
Campbell High School           407          407            407  
Campos High School             541          541            541  
Chang High School              171          171            171  
Dixon High School             1583         1583           1583  
Fisher High School             798          798            798  
Green High School             1961         1961           1961  
Montgomery High School        2038         2038           2038  
Odonnell High School           459          459            459  
Richard High School            551          551            551  
Silva High School             1109         1109           1109  
Sullivan High School           971          971            971  
Turner High School             846          846            846  
Wagner High School            1541         1541           1541  
Woods High School             1052         1052           1052

In [65]:
sorted_ascending = total_students_by_school.sort_values(by = ["student_id"])
sorted_ascending

student_id  student_name  grade  reading_score  \
school_name                                                              
Chang High School              171           171    171            171   
Campbell High School           407           407    407            407   
Odonnell High School           459           459    459            459   
Campos High School             541           541    541            541   
Richard High School            551           551    551            551   
Fisher High School             798           798    798            798   
Bowers High School             803           803    803            803   
Turner High School             846           846    846            846   
Sullivan High School           971           971    971            971   
Woods High School             1052          1052   1052           1052   
Silva High School             1109          1109   1109           1109   
Wagner High School            1541          1541   1541           1541   
Dixon High School             1583          1583   1583           1583   
Green High School             1961          1961   1961           1961   
Montgomery High School        2038          2038   2038           2038   

                        math_score  school_type  school_budget  
school_name                                                     
Chang High School              171          171            171  
Campbell High School           407          407            407  
Odonnell High School           459          459            459  
Campos High School             541          541            541  
Richard High School            551          551            551  
Fisher High School             798          798            798  
Bowers High School             803          803            803  
Turner High School             846          846            846  
Sullivan High School           971          971            971  
Woods High School             1052         1052           1052  
Silva High School             1109         1109           1109  
Wagner High School            1541         1541           1541  
Dixon High School             1583         1583           1583  
Green High School             1961         1961           1961  
Montgomery High School        2038         2038           2038

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

This dataset contained student data on reading and math scores based on school budget, school type and grade of the student. 

There was some cleaning to do as there were some null values in the reading and math scores. It makes sense to remove these values since if the students didn't take the test, then we don't need their names in the analysis. When we got rid of duplicate rows, we perhaps could've taken the average of the students 2 scores, if they took the test twice. 

We were able to calculate the average math score, which was 65, although this doesn't mean much without making a comparison between subsets of data. 

For grade 9 students, the average reading score is 69 and the average math score is 67. However it would be interesting to make this comparison across different schools and school types. 

We were able to determine that the lowest reading score was 10.5. This student was named Matthew Thomas from Dixon High School. 

The average reading score for grade 11 and 12 students was 75. So grade 11 and 12 students seem to score higher on average compared to grade 9 students in reading (who scored an average of 69 in reading). 

Lastly, we were able to make a comparison between public and charter schools. On average, charter school students score 72 and 67 in reading and math respectively. Public school students score 72 and 63 in reading and math respectively. So it seems that charter school students score slightly higher in math, but there are no difference in reading scores. 

It would be interesting to compare reading and math scores by grade, by school type, school budget, etc and do a more in depth analysis. 